### ISSUE TITLE
[Feature Request] Provide documentation for testing with Amplify

### ISSUE THREAD

**Issue by Jordan-Nelson** [aws-amplify/amplify-flutter#1256]
### Context

There have been a few issues opened asking questions about testing flutter apps with Amplify (unit and integration). The amplify flutter docs currently don't offer much guidance on testing. 

### Request

At a minimum:
- Add basic documentation for writing unit tests and integration tests with Amplify Flutter. A simple example and link to the flutter docs may be sufficient.
- Callout any known limitations (hosted UI)

Ideally:
- Provide info on writing integration tests with hosted UI. Currently this is a limitation, and a work around would need to be identified. See: https://github.com/aws-amplify/amplify-flutter/issues/950
- Provide suggested solutions to some of the challenging aspects of writing integration tests with Amplify (confirmation codes) that we have already solved for our own integration testing needs.

### Related issues
- https://github.com/aws-amplify/amplify-flutter/issues/950
- https://github.com/aws-amplify/amplify-flutter/issues/1255

**Comment by 94Sip** [comment]
did this documentation ever get created? If so, can you point me to it? I'd like to understand how to do integration testing when using the Amplify Authenticator component and other Amplify libraries needing Auth (storage and API, in particular).

Thank you.

**Comment by Jordan-Nelson** [comment]
@94Sip no, I don't believe there are any public docs on integration or end to end testing. Are you looking to write tests that make network calls to AWS services (a full end to end test), or are you looking to stub/mock some portion to test your application logic?

**Comment by 94Sip** [comment]
I am looking to do end-to-end testing - I am using the Authenticator UI component, and all my API calls are AWS_IAM authorization.

**Comment by Jordan-Nelson** [comment]
Got it. Is there a particular challenge you are facing with the tests? One common problem with Auth is creating new users, as by default that requires sending a confirmation code either via phone or email.

**Comment by 94Sip** [comment]
nothing in general, but I would like to see how it is done. Meaning, do I just load my app (just like all the integration test scenarios show), let Amplify configure() happen and then wait for pump and settle for the Authenticator to show the login form, and then proceed as usual? I guess just seeing some working code, or even pseudo code for this type of scenario would be helpful. Do I need to do anything so that the IAM authorization works as expected in an integration testing setup?

**Comment by 94Sip** [comment]
Update @Jordan-Nelson - I was able to get it to work!

After typing in my thoughts above, I figured why not just try it out, and sure enough, it worked.

Caveat: I did not create a integration test for creating new users.

**Comment by Jordan-Nelson** [comment]
Got it. Most of the widgets in the Authenticator have [specific Keys](https://github.com/aws-amplify/amplify-flutter/blob/main/packages/authenticator/amplify_authenticator/lib/src/keys.dart), which we use for our own widget tests. These are not part of the public API and could change in the future, but could be helpful for testing.

**Comment by 94Sip** [comment]
yes! I looked thru Widget explored to identify those keys - i didn't realize they were documented in source, as you linked.

Question @Jordan-Nelson - back to one of your previous questions: how do I integration test user signup where a confirmation code is sent by Cognito?

**Comment by Jordan-Nelson** [comment]
@94Sip there are a couple options you could use in a testing environment. You would not want to use any of these options in a production environment.
1. Automatically confirm the user, removing the need for confirmation via code. See [this post](https://repost.aws/knowledge-center/cognito-automatically-confirm-users) for an example of how to do that. This is probably the easiest option if you aren't too concerned about including the confirmation flow in your test.
2. Use a Cognito lambda trigger to capture and store the confirmation code. Cognito offers triggers (lambdas) that can be set up to run on many different events. One of those events is a Custom Email Sender or a Customer SMS Sender. These lambda triggers will run each time Cognito would normally send a email or SMS. You can use this to intercept the code and store it somewhere (for example AppSync) and then your automated test can fetch it. This is how we test the full sign up process in our end to end tests in Amplify. Setting this up is fairly involved. I would only recommend this if you had a good reason to test this flow as part of your e2e tests. You can see how we use the AWS CDK to set up these triggers in our e2e tests here: https://github.com/aws-amplify/amplify-flutter/tree/main/infra-gen2/infra-common/src/auth-user-extensions. That directory has some other utils for creating and deleting users as well.

**Comment by 94Sip** [comment]
Extremely helpful @Jordan-Nelson thank you!
